cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/monitor/sampling/aggregator/KafkaMetricSampleAggregator.java

```diff
@@ -0,0 +1,290 @@
+   * <ol>
+   *   <li>
+   *     We assume that only new topics would have insufficient data. So we only balance the existing topics and
+   *     allow more time to collect enough utilization data for the new topics.
+   *   </li>
+   *   <li>
+   *     If we don't have enough data to make a replica movement decision, it is better not to take any action.
+   *   </li>
+   * </ol>
+   *
+   * @param clusterAndGeneration The current cluster information.
+   * @param now the current time.
+   * @param operationProgress to report the async operation progress.
+   * @return The {@link MetricSampleAggregationResult} for all the partitions.
+   */
+  public MetricSampleAggregationResult<String, PartitionEntity> aggregate(MetadataClient.ClusterAndGeneration clusterAndGeneration,
+                                                                          long now,
+                                                                          OperationProgress operationProgress)
+      throws NotEnoughValidWindowsException {
+    ModelCompletenessRequirements requirements = new ModelCompletenessRequirements(1, 0.0, false);
+    return aggregate(clusterAndGeneration, -1L, now, requirements, operationProgress);
+  }
+
+  /**
+   * Collect the aggregated metrics for all the topic partitions for a time window.
+   * <p>
+   * If a topic has at least one window that does not have enough samples, that topic will be excluded from the
+   * returned aggregated metrics. This is because:
+   * <ol>
+   *   <li>
+   *     We assume that only new topics would have insufficient data. So we only balance the existing topics and
+   *     allow more time to collect enough utilization data for the new topics.
+   *   </li>
+   *   <li>
+   *     If we don't have enough data to make a replica movement decision, it is better not to take any action.
+   *   </li>
+   * </ol>
+   *
+   * @param clusterAndGeneration The current cluster information.
+   * @param from the start of the time window
+   * @param to the end of the time window
+   * @param requirements the {@link ModelCompletenessRequirements} for the aggregation result.
+   * @param operationProgress to report the operation progress.
+   * @return The {@link MetricSampleAggregationResult} for all the partitions.
+   */
+  public MetricSampleAggregationResult<String, PartitionEntity> aggregate(MetadataClient.ClusterAndGeneration clusterAndGeneration,
+                                                                          long from,
+                                                                          long to,
+                                                                          ModelCompletenessRequirements requirements,
+                                                                          OperationProgress operationProgress)
+      throws NotEnoughValidWindowsException {
+    RetrievingMetrics step = new RetrievingMetrics();
+    try {
+      operationProgress.addStep(step);
```

Comment:
Should we check for the case where `operationProgress` could be `null`?

Reply:
Our check is indeed a little loose here primarily because we control the input as well. We know that the `clusterGeneration`, `requirements` and `operationProgress` will not be null.